contracts/AddressBook.sol

```diff
@@ -3,10 +3,119 @@
+    }
+
+    function getOtokenFactory() external view returns (address) {
+        return getAddress(OTOKEN_FACTORY);
+    }
+
+    function getWhitelist() external view returns (address) {
+        return getAddress(MARGIN_WHITELIST);
+    }
+
+    function getController() external view returns (address) {
+        return getAddress(CONTROLLER);
+    }
+
+    function getMarginPool() external view returns (address) {
+        return getAddress(MARGIN_POOL);
+    }
+
+    function getMarginCalculator() external view returns (address) {
+        return getAddress(MARGIN_CALCULATOR);
+    }
+
+    function getLiquidationManager() external view returns (address) {
+        return getAddress(LIQUIDATION_MANAGER);
+    }
+
+    function getOracle() external view returns (address) {
+        return getAddress(ORACLE);
+    }
+
+    function setAddress(bytes32 _key, address _module) public onlyOwner {
+        modules[_key] = _module;
+    }
+
+    function setOtokenImpl(address _otokenImpl) external onlyOwner {
+        updateImplInternal(OTOKEN_IMPL, _otokenImpl);
+    }
+
+    function setOtokenFactory(address _otokenFactory) external onlyOwner {
+        updateImplInternal(OTOKEN_FACTORY, _otokenFactory);
+    }
+
+    function setWhitelist(address _whitelist) external onlyOwner {
+        updateImplInternal(MARGIN_WHITELIST, _whitelist);
+    }
+
+    function setController(address _controller) external onlyOwner {
+        updateImplInternal(CONTROLLER, _controller);
+    }
+
+    function setMarginPool(address _marginPool) external onlyOwner {
+        updateImplInternal(MARGIN_POOL, _marginPool);
+    }
+
+    function setMarginCalculator(address _marginCalculator) external onlyOwner {
+        updateImplInternal(MARGIN_CALCULATOR, _marginCalculator);
+    }
+
+    function setLiquidationManager(address _liquidationManager) external onlyOwner {
+        updateImplInternal(LIQUIDATION_MANAGER, _liquidationManager);
+    }
+
+    function setOracle(address _oracle) external onlyOwner {
+        updateImplInternal(ORACLE, _oracle);
+    }
+
+    /**
+     * @dev internal function to update the implementation of a specific component of the protocol
+     * @param _id the id of the contract to be updated
+     * @param _newAddress the address of the new implementation
+     **/
+    function updateImplInternal(bytes32 _id, address _newAddress) public onlyOwner {
+        address payable proxyAddress = address(uint160(getAddress(_id)));
+
+        OwnedUpgradeabilityProxy proxy = OwnedUpgradeabilityProxy(proxyAddress);
+        bytes memory params = abi.encodeWithSignature("initialize(address)", address(this));
 
+        if (proxyAddress == address(0)) {
+            proxy = new OwnedUpgradeabilityProxy();
+            proxy.upgradeToAndCall(_newAddress, params);
+            setAddress(_id, address(proxy));
+            emit ProxyCreated(_id, address(proxy));
+        } else {
+            proxy.upgradeToAndCall(_newAddress, params);
```

Comment:
Refactor to

```javascript
address payable proxyAddress = address(uint160(getAddress(_id)));
bytes memory params = abi.encodeWithSignature("initialize(address)", address(this));
if (proxyAddress == address(0)) {
     OwnedUpgradeabilityProxy proxy = new OwnedUpgradeabilityProxy();
     proxy.upgradeToAndCall(_newAddress, params);
     setAddress(_id, address(proxy));
     emit ProxyCreated(_id, address(proxy));
} else {
       OwnedUpgradeabilityProxy proxy = OwnedUpgradeabilityProxy(proxyAddress);
       proxy.upgradeToAndCall(_newAddress, params);
 } 
```

Reply:
it looks like the written code is a refactor of ur comment ?!

Reply:
in the `if` block, you declare a new proxy instance, 
so you should just put this line
`OwnedUpgradeabilityProxy proxy = OwnedUpgradeabilityProxy(proxyAddress);` in the else block?

Reply:
@haythem96 thoughts?

Reply:
the line u added in the `else` block is the L109 in the current implementation, for me the current implementation looks like a refactor for what u wrote.
Also is this so important ? does it make an issue in term of execution ?

Reply:
Yeah I'm suggesting putting that line in the else block.

Reply:
I just feel like you don't need L109 outside of the if-else condition, because if (proxyAddress==0), you'll just waste the gas that executes L109 right?

Reply:
okay I changed that.